File: Caesar Cipher.c
#include <stdio.h>
#include <string.h>
#include <ctype.h>
#include <stdlib.h>

// This initial calling of functions is needed because
// the following functions were declared after the main function.
void print_usage();
void encrypt();
void decrypt();
char *lowercaseString(char string[]);

// The parameters in main function are for accepting command line arguments (CLAs).
int main(int argc, char *argv[])
{
    // This if statement exits the program if there are more than two CLAs.
    // The first of those two CLAs, argv[0], is the script name.
    if (argc != 2)
    {
        print_usage();
    }

    // The strcmp function takes two strings as arguments.
    // If the strings are equal, return the number 0.
    // This function must be used in the conditional statement instead of: if (argv[1] == "--encrypt")
    if (strcmp(argv[1], "--encrypt") == 0)
    {
        encrypt();
    }
    else if (strcmp(argv[1], "--decrypt") == 0)
    {
        decrypt();
    }
    else
    {
        print_usage();
    }

    return 0;
}

void encrypt()
{
    char plaintext[256];
    printf("\nEnter the message you want to encrypt: ");
    fgets(plaintext, 256, stdin);

    lowercaseString(plaintext);

    char line[256];
    int shift;
    printf("Enter the desired shift value (integer between 1 and 25): ");
    fgets(line, sizeof(line), stdin);
    sscanf(line, "%d", &shift);

    if (shift > 25 || shift < 0)
    {
        printf("Invalid shift value.\n\n");
        exit(1);
    }

    printf("%d. ", shift);
    for (int j = 0; j < strlen(plaintext); ++j)
    {
        char letter = plaintext[j] - 97;

        if (letter > 25 || letter < 0)
        {
            printf("%c", plaintext[j]);
            continue;
        }

        int formula = (letter + shift) % 26;

        if (formula > 25)
        {
            formula -= 26;
        }
        else if (formula < 0)
        {
            formula += 26;
        }

        letter = formula + 97;

        printf("%c", letter);
    }

    printf("\n\n");
}

void decrypt()
{
    char ciphertext[256];
    printf("\nEnter the Caesar Cipher encoded text: ");
    fgets(ciphertext, 256, stdin);

    lowercaseString(ciphertext);

    for (int shift = 1; shift != 26; ++shift)
    {
        printf("%d. ", shift);
        for (int j = 0; j < strlen(ciphertext); ++j)
        {
            char letter = ciphertext[j] - 97;

            if (letter > 25 || letter < 0)
            {
                printf("%c", ciphertext[j]);
                continue;
            }

            int formula = (letter - shift) % 26;

            if (formula > 25)
            {
                formula -= 26;
            }
            else if (formula < 0)
            {
                formula += 26;
            }

            letter = formula + 97;

            printf("%c", letter);
        }
    }

    printf("\n\n");
}

void print_usage()
{
    printf("\nIncorrect argument usage.\n");
    printf("Run either `./Caesar_Cipher --encrypt` or `./Caesar_Cipher --decrypt`.\n\n");

    // The exit() function will exit the program if called anywhere in the code.
    // A return statement from within the main function will act in the same way.
    // The exit() function requires an integer as its parameter.
    // No matter the integer value, the exit() function will exit the program.
    // That integer is the exit status. Exit status 1 is "Operation not permitted".
    exit(1);
}

// return type is a pointer to the first element in the string
char *lowercaseString(char string[])
{
    for (int i = 0; i != strlen(string); ++i)
    {
        string[i] = tolower(string[i]);
    }

    return string;
}

File: Caesar Cipher.py
#!/usr/bin/env python3

import random
import sys


def encrypt():
    # getting the ciphertext and making it lowercase
    plaintext = input("Enter the plaintext to encode: ").lower()
    print("")

    shift_type = input("Random or custom shift? (r/c): ")
    if shift_type == "r":
        shift = random.randint(1, 25)
    elif shift_type == "c":
        shift = int(input("Enter the shift to encode (1 - 25): "))

        if shift > 25 or shift < 1:
            print("Invalid shift.")
            sys.exit()
    else:
        print("Invalid input.")
        sys.exit()

    # start: variable shift = 1
    # stop: variable shift = 26
    print(f"{shift}. ", end="")
    for char in plaintext:
        # ord() gets the ASCII code of a character
        # ascii value of lowercase letter - 97 makes:
        # 'a' = 0, 'b' = 1, ... 'z' = 25

        letter = ord(char) - 97

        if letter > 25 or letter < 0:
            # used to account for symbols
            # prints whatever character isn't in the alphabet
            print(char, end="")
            continue  # go back to top of body of loop

        # defining the formula for the Caesar Cipher
        formula = (letter + shift) % 26

        if formula > 25:
            formula -= 26
        elif formula < 0:
            formula += 26

        # chr() gets the character encoded by an ASCII code number
        # the + 97 is used to make 'a' go from 0 to 97
        # (back to its ascii value)
        letter = chr(formula + 97)

        print(letter, end="")

    print("")

    # this print statement is to end with a newline
    print("")


def decrypt():
    # getting the ciphertext and making it lowercase
    ciphertext = input("Enter the Caesar Cipher encoded text: ").lower()
    print("")

    for shift in range(1, 26):
        # start: variable shift = 1
        # stop: variable shift = 26
        print(f"{shift}. ", end="")
        for char in ciphertext:
            # ord() gets the ASCII code of a character
            # ascii value of lowercase letter - 97 makes:
            # 'a' = 0, 'b' = 1, ... 'z' = 25

            letter = ord(char) - 97

            if letter > 25 or letter < 0:
                print(char, end="")
                continue  # go back to top of body of loop

            # defining the formula for the Caesar Cipher
            formula = (letter - shift) % 26

            if formula > 25:
                formula -= 26
            elif formula < 0:
                formula += 26

            # chr() gets the character encoded by an ASCII code number
            # the + 97 is used to make 'a' go from 0 to 97
            # (back to its ascii value)
            letter = chr(formula + 97)

            print(letter, end="")

        shift += 1
        print("")

    # this print statement is to end with a newline
    print("")


def main():
    # args is a list of the command line arguments
    # sys.argv[0] is the script name
    # [0:] starts splicing at the zero index to getting all indexes from 0+

    args = sys.argv[0:]
    if len(args) != 2:
        print("Incorrect argument usage.")
        sys.exit()
    else:
        arg = args[1]

    if arg == "--encrypt" or arg == "-e":
        encrypt()
    elif arg == "--decrypt" or arg == "-d":
        decrypt()
    else:
        print("Incorrect argument usage.")
        sys.exit()


if __name__ == '__main__':
    main()
